Add WASM bindings using wasm-bindgen

jp-knj opened the pull request:
### Overview                                                                                                                                                                                                                                      

Provides minimal WebAssembly bindings for markdown-rs, addressing the WASM hanging issues mentioned in #182 while following maintainer preference for WASM-only approach (no NAPI). 

### Background                                                                                                                                                                                                                                   

This PR takes a different approach from #182                                                                                                                                                                                               
- Prev PR: NAPI + WASM hybrid with platform-specific builds
- This PR: Pure WASM solution, simpler and more maintainable                                                                                                                                                                                                                                                                                                                           

#### Key Improvements                                                                                                                                                                                                                              

- Fixes WASM hanging - Resolves the hanging issue reported in #182
- No NAPI complexity - Pure WASM
- ESM-first - Uses `.mjs` extension for proper Node.js module resolution
- Minimal footprint - Single wasm binary (~500KB) vs multiple platform builds

### Implementation                                                                                                                                                                                                                               

```js
import { toHtml, toHtmlWithOptions } from 'markdown-wasm';

// CommonMark
const html = await toHtml('# Hello World');

// With GFM/MDX
const gfmHtml = await toHtmlWithOptions('~text~', { gfm: true });
```

#### Why Pure WASM?                                                                                                                                                                                                                               

Per maintainer feedback on #182:
- Lower maintenance burden (no per-platform builds)
- Universal compatibility (one binary for all)
- No native toolchain requirements
- Simpler CI/CD pipeline

### Testing                                                                                                                                                                                                                                      

- Works on Node.js 16+, Bun 1.1+
- No hanging issues
- Comprehensive test suite included
- Examples for basic and advanced usage

### Performance

While NAPI may be faster, WASM provides:
- Consistent performance across platforms
- No native compilation for users
- Adequate speed for most use cases

This approach prioritizes maintainability and reliability over maximum performance, aligning with project goals.                                                                                                                             

Related to #182

jp-knj committed:
feat(wasm): add WebAssembly bindings

- Add wasm-bindgen setup with Cargo.toml
- Implement to_html and to_html_with_options APIs
- Support CommonMark, GFM, MDX, and frontmatter

jp-knj committed:
feat(wasm): add Node.js ES module wrapper

- Create async/sync JavaScript API wrapper
- Configure package.json with .mjs build script
- Handle WASM initialization and module loading

jp-knj committed:
test(wasm): add comprehensive test suite

- Test basic markdown conversions
- Test GFM, MDX, and security options
- Test sync/async APIs
- Cover edge cases and complex markdown

jp-knj committed:
docs(wasm): add README, examples, and LICENSE

- Add comprehensive README with API documentation
- Add basic usage example
- Add advanced options example
- Include MIT LICENSE file

jp-knj committed:
build: configure monorepo for WASM package

- Add wasm to Cargo workspace
- Setup pnpm workspace with wasm directory
- Add root package.json for development scripts
- Update .gitignore for WASM build artifacts

wooorm commented:
Thanks for working on this!
Could we do with plain npm?
should it be called `markdown-rs-wasm`, what about `markdown-wasm`? Something else?

jp-knj committed:
feat(wasm): prepare @wooorm/markdown-wasm for plain-npm publish

- Scope package as @wooorm/markdown-wasm
- Export lib/index.mjs (+ types)
- Add examples script and require Node >=18
- Update Whitelist files

Refs: #185

jp-knj commented:
I’m not fully confident in this change.
does it look good to you? Happy to adjust if needed.

ChristianMurphy commented:
I haven't had a chance to pull down the code and try it out (traveling presently).
But at an approach/strategy level, I'm a lot more comfortable with a single WASM build, over half a dozen single platform builds with N-API bindings.

jp-knj committed:
feat(docs): add WASM package documentation and references

- Add @wooorm/markdown-wasm
- Update wasm/README.md with correct package name
- Fix installation and import examples to use @wooorm/markdown-wasm

punkpeye commented:
Is the AST compatible with `mdast-util-from-markdown` ? If so, it would be really nice to expose that.
